code/go/0chain.net/blobbercore/allocation/protocol.go

```diff
@@ -52,16 +54,31 @@ func (a *Allocation) LoadTerms(ctx context.Context) (err error) {
 func FetchAllocationFromEventsDB(ctx context.Context, allocationID string, allocationTx string, readonly bool) (a *Allocation, err error) {
 	var tx = datastore.GetStore().GetTransaction(ctx)
 
-	a = new(Allocation)
-	err = tx.Model(&Allocation{}).
-		Where(&Allocation{Tx: allocationTx}).
-		First(a).Error
+	isAllocationUpdated := false
+	isAllocationDetailsCached := false
 
-	if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
-		return nil, common.NewError("bad_db_operation", err.Error()) // unexpected DB error
+	cachedAllocationInterface, err := lru.Get(allocationID)
+	if err == nil {
+		isAllocationDetailsCached = true
+		cachedAllocation := cachedAllocationInterface.(*Allocation)
+
+		if a.Tx == allocationTx {
+			a = cachedAllocation
+		} else {
+			isAllocationUpdated = true
+		}
+	} else {
+		a = new(Allocation)
+		err = tx.Model(&Allocation{}).
+			Where(&Allocation{Tx: allocationTx}).
+			First(a).Error
+
+		if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
```

Comment:
wha if it's not in db? we should go to sharders

Reply:
If it is not in DB we have requestAllocation below it. So if it exists we just return it and if it doesn't then we fetch it and add to db and cache.